File: swift/Tests/FundamentalCoreTests/ParticleShapeTests.swift
import Testing
import Foundation
import FundamentalCore

// `simd_length` is an Apple-platform C function; use a plain Pythagorean length so tests
// compile on Linux where the simd C bridge isn't available.
private func len(_ v: SIMD2<Float>) -> Float { Foundation.sqrt(v.x * v.x + v.y * v.y) }

@Suite("ParticleShape")
struct ParticleShapeTests {

    @Test("dot is the fast circle — no polygon vertices")
    func dot() {
        #expect(ParticleShape.dot.vertices == nil)
        #expect(ParticleShape.dot.isDot)
    }

    @Test("star yields 2×points vertices, first straight up, alternating outer/inner radius")
    func star() {
        let s = ParticleShape.star(points: 5, innerRatio: 0.5)
        let v = try! #require(s.vertices)
        #expect(v.count == 10)
        #expect(!s.isDot)
        // first point straight up (screen y-down): (0, -1) at the outer radius.
        #expect(abs(v[0].x) < 1e-5)
        #expect(abs(v[0].y + 1) < 1e-5)
        // even indices ride the outer radius (1), odd indices the inner radius (innerRatio).
        #expect(abs(len(v[2]) - 1) < 1e-5)
        #expect(abs(len(v[1]) - 0.5) < 1e-5)
    }

    @Test("polygon yields N unit-circle vertices")
    func polygon() {
        #expect(ParticleShape.polygon(sides: 3).vertices?.count == 3)
        let hex = try! #require(ParticleShape.polygon(sides: 6).vertices)
        #expect(hex.count == 6)
        for vertex in hex { #expect(abs(len(vertex) - 1) < 1e-5) }
    }

    @Test("custom carries the given vertices verbatim")
    func custom() {
        let verts: [SIMD2<Float>] = [SIMD2(0, -1), SIMD2(1, 1), SIMD2(-1, 1)]
        #expect(ParticleShape.custom(verts).vertices == verts)
        #expect(!ParticleShape.custom(verts).isDot)
    }

    @Test("a render frame carries the configured shape through to the renderer")
    func framePropagatesShape() {
        // FieldOptions → RenderFrame is the channel the CoreGraphics / Metal renderers read.
        let opts = FieldOptions(render: .dots, particleShape: .star(points: 6))
        #expect(opts.particleShape.vertices?.count == 12)
    }
}

File: android/fundamental-core/src/test/kotlin/com/fundamental/core/ParticleShapeTests.kt
package com.fundamental.core

import com.fundamental.core.engine.ParticleShape
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

// ParticleShape (§render) — unit vector shapes the host stamps per particle. Ported from Swift; the
// vertex math is checked exactly.

class ParticleShapeTests {

    @Test
    fun dotIsTheFastCirclePath() {
        assertTrue(ParticleShape.DOT.isDot)
        assertNull(ParticleShape.DOT.vertices)
    }

    @Test
    fun starHasTwoVerticesPerPointAlternatingRadius() {
        val s = ParticleShape.star(points = 5, innerRatio = 0.5f)
        val v = s.vertices!!
        assertEquals(10, v.size, "5-point star → 10 vertices (outer+inner)")
        // even indices are the outer points (radius 1), odd are the waist (radius 0.5)
        assertEquals(1f, kotlin.math.hypot(v[0].x, v[0].y), 1e-4f)
        assertEquals(0.5f, kotlin.math.hypot(v[1].x, v[1].y), 1e-4f)
        // first point straight up: angle −π/2 → (0, −1)
        assertTrue(kotlin.math.abs(v[0].x) < 1e-4f && v[0].y < 0f, "first point is straight up")
        assertTrue(!s.isDot)
    }

    @Test
    fun polygonIsNUnitVertices() {
        val tri = ParticleShape.polygon(sides = 3)
        assertEquals(3, tri.vertices!!.size)
        tri.vertices!!.forEach { assertEquals(1f, kotlin.math.hypot(it.x, it.y), 1e-4f, "polygon vertices are unit-radius") }
        // sides clamps to ≥ 3
        assertEquals(3, ParticleShape.polygon(sides = 1).vertices!!.size)
        assertEquals(6, ParticleShape.polygon(sides = 6).vertices!!.size)
    }

    @Test
    fun customCarriesTheGivenVertices() {
        val verts = listOf(com.fundamental.core.math.Vec3(0f, -1f, 0f), com.fundamental.core.math.Vec3(1f, 1f, 0f), com.fundamental.core.math.Vec3(-1f, 1f, 0f))
        assertEquals(verts, ParticleShape.custom(verts).vertices)
    }
}
